### ISSUE TITLE
"fallthrough" incorrectly supported as a function attribute

### ISSUE THREAD

**Issue by kusma** [mesonbuild/meson#8021]
51e9db370a0ebccaf220e171c3444a0f2c4e1723 added support for `has_function_attribute()`, and it included `fallthrough` as a supported value.

But that's not a [function attribute](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes) at all, it's a [statement attribute](https://gcc.gnu.org/onlinedocs/gcc/Statement-Attributes.html#Statement-Attributes).

There's no support for querying support for statement-attributes, though.

It seems there's reasonable-ish ways of resolving this:
1. Remove `fallthrough` from the documentation, and emit a warning if used.
2. Add API for querying statement-attributes, emit a warning if queried for the wrong type
3. Rename API to `compiler.get_supported_attributes()`, and deprecate the old API

Option 2 seems a bit silly, as this is the *only* supported statement-attribute according to the GCC docs at the time of writing.

**Comment by kusma** [comment]
@dcbaker: Thoughts? It seems you introduced this in the first place. What was your original intention here?

**Comment by dcbaker** [comment]
I'm pretty sure I just copied the list from autotools, so they probably had this in there. I've meant to add a `statement-attributes`, but haven't gotten there yet.

**Comment by dcbaker** [comment]
@kusma If there's something specific we need in mesa let me know, and I'll short list that.

**Comment by kusma** [comment]
@dcbaker: I don't think we need anything, `__has_attribute()` seems to do the trick for us. I just discovered this when looking into how to do this similar to how we do function-attribute stuff, so I thought I would report this mistake. See [this MR](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/7747) for the use-case.

**Comment by gsauthof** [comment]
@kusma [wrote](https://github.com/mesonbuild/meson/issues/8021#issue-749634703):

> Option 2 seems a bit silly, as this is the _only_ supported statement-attribute according to the GCC docs at the time of writing.

FWIW, nowadays GCC [supports at least](https://gcc.gnu.org/onlinedocs/gcc/Common-Attributes.html) the following statement attributes:

- assume
- hot
- cold
- fallthrough


Somewhat related, `[[fallthrough]]` and `[[ __fallthrough__ ]] ` is part of the [C standard since C23](https://en.cppreference.com/w/c/language/attributes/fallthrough.html) (and the former is also part of C++ since C++17). It isn't entirely obvious to me how to test for such standardized attributes, with meson.
Or query the C standard version the compiler supports.

**Comment by gsauthof** [comment]
This is the best I came up with:

```
if cc.get_define('__has_c_attribute(fallthrough)') != '0'
    # available since C23
    conf.set('FALLTHROUGH', '[[fallthrough]]')
elif cc.has_function_attribute('fallthrough')
    # actually, falltrhough is a statement attribute,
    # but work-around until https://github.com/mesonbuild/meson/issues/8021 is fixed
    conf.set('FALLTHROUGH', '__attribute__((fallthrough))')
else
    conf.set('FALLTHROUGH', '')
endif
```
